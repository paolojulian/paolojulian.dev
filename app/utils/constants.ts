/**
 * These are the links to the different apps
 */
export const ExternalLinks = {
  articles: process.env.NEXT_PUBLIC_ARTICLES_URL ?? 'http://localhost:3002',
  oldApp: process.env.NEXT_PUBLIC_OLD_APP_URL ?? '',
};

export const Routes = {
  Home: '/',
  About: '/about-me',
  Apps: '/apps',
  Articles: `${ExternalLinks.oldApp}/blogs`,
  Contact: `${ExternalLinks.oldApp}/contact`,
} satisfies Record<string, string>;

export const DynamicRoutes = {
  ArticleItem: (slug: string) => `${ExternalLinks.articles}/blogs/${slug}`,
};
