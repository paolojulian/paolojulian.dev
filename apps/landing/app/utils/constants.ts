/**
 * These are the links to the different apps
 */
export const ExternalLinks = {
  articles: process.env.NEXT_PUBLIC_ARTICLES_URL ?? 'http://localhost:3002',
};

export const Routes = {
  Home: '/',
  About: '/about-me',
  Apps: '/apps',
  Articles: ExternalLinks.articles,
  Contact: '/contact',
} satisfies Record<string, string>;
