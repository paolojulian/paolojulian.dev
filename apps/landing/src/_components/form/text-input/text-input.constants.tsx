import classNames from 'classnames';

export type Variants = 'default' | 'default-dark';

export const COLOR_VARIANTS: Record<Variants, string> = {
  default: classNames(
    'border-slate-400 focus:border-slate-800', // border
    'text-slate-700' // font
  ),
  'default-dark': classNames(
    'border-slate-600 focus:border-slate-50', // border
    'text-slate-200 placeholder-slate-500' // font
  ),
};

export const DATA_TEST = {
  container: 'text-input-container',
  input: 'text-input-input',
  label: 'text-input-label',
};
