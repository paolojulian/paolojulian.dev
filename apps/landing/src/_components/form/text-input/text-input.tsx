import { DATA_TEST } from '@/_components/form/text-input/text-input.constants';
import classNames from 'classnames';
import React, {
  FunctionComponent,
  InputHTMLAttributes,
  useMemo,
  useState,
} from 'react';

type Variants = 'default' | 'default-dark';

const colorVariants: Record<Variants, string> = {
  default: 'border-new-highlight text-new-white placeholder-new-highlight',
  'default-dark':
    'border-new-highlight focus:border-new-white text-new-white placeholder-new-highlight',
};

interface Props extends InputHTMLAttributes<HTMLInputElement> {
  id: string;
  label: string;
  name: string;
  variant?: Variants;
  isError?: boolean;
  testId?: string;
}

const TextInput: FunctionComponent<Props> = ({
  variant = 'default',
  testId = DATA_TEST.input,
  id,
  isError,
  label,
  name,
  value,
  onFocus = () => {},
  onBlur = () => {},
  ...props
}) => {
  const [isFocused, setIsFocused] = useState(!!value || props.autoFocus);

  const isFocusedComputed = useMemo(
    () => !!value || isFocused,
    [value, isFocused]
  );

  const handleFocus = (e: React.FocusEvent<HTMLInputElement>) => {
    onFocus(e);
    setIsFocused(true);
  };

  const handleBlur = (e: React.FocusEvent<HTMLInputElement>) => {
    onBlur(e);
    setIsFocused(false);
  };

  return (
    <div
      className={classNames(
        'relative w-full group',
        'transition-colors bg-inherit border-b',
        isFocusedComputed
          ? 'border-opacity-100'
          : 'border-opacity-20 hover:border-opacity-50',
        isError ? 'border-red-500' : colorVariants[variant]
      )}
      data-testid={DATA_TEST.container}
    >
      {/* Top label (When text input is in focused state) */}
      <label
        className={classNames(
          'absolute left-3 top-0',
          'transition duration-200 pointer-events-none',
          'font-medium text-sm text-new-highlight',
          isFocusedComputed ? 'scale-100 opacity-100' : 'scale-0 opacity-0'
        )}
        htmlFor={id}
        data-testid={DATA_TEST.label}
      >
        {label}
      </label>

      {/* Placeholder label (When textinput is not in focused state) */}
      <div
        className={classNames(
          'absolute bottom-2 left-2',
          'transition duration-200 pointer-events-none',
          'text-lg text-new-highlight',
          isFocusedComputed ? 'scale-0 opacity-0' : 'scale-100 opacity-100'
        )}
      >
        {label}
      </div>

      <input
        {...props}
        id={id}
        name={name}
        value={value}
        className={classNames(
          'w-full px-3 pt-6 pb-2',
          'bg-inherit',
          'focus:outline-none transition-opacity text-new-white',
          isFocusedComputed ? 'opacity-100' : 'opacity-0'
        )}
        onFocus={handleFocus}
        onBlur={handleBlur}
        data-testid={testId}
      />
    </div>
  );
};

export default TextInput;
