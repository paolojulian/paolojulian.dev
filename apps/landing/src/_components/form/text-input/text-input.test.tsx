import { DATA_TEST } from './text-input.constants';
import TextInput from './text-input';
import { render, screen } from '@testing-library/react';

type TextInputProps = React.ComponentProps<typeof TextInput>;

describe('TESTING TextInput Component', () => {
  describe('GIVEN a TextInput component with variant and isError prop', () => {
    const defaultProps: TextInputProps = {
      variant: 'default',
      isError: false,
      id: 'testing',
      name: 'testing',
      label: 'testing',
    };

    function renderComponent(props: TextInputProps) {
      return render(<TextInput {...props} />);
    }

    describe('WHEN the TextInput Component is rendered on snapshot', () => {
      describe('THEN it should match the snapshot', () => {
        const { asFragment } = renderComponent(defaultProps);
        expect(asFragment).toMatchSnapshot();
      });
    });

    describe('WHEN the TextInput Component is rendered', () => {
      beforeEach(() => {
        renderComponent(defaultProps);
      });

      it('THEN it should match the snapshot', () => {
        const { asFragment } = render(<TextInput {...defaultProps} />);
        expect(asFragment()).toMatchSnapshot();
      });

      it('THEN it should render the component', () => {
        const input = screen.getByTestId(DATA_TEST.container);
        expect(input).toBeInTheDocument();
      });

      it('THEN it should not apply error styles', () => {
        const input = screen.getByTestId(DATA_TEST.container);
        expect(input).not.toHaveClass('border-red-500');
      });
    });

    describe('WHEN the TextInput Component has isError set to true', () => {
      const props = { ...defaultProps, isError: true };
      beforeEach(() => {
        renderComponent({ ...props });
      });

      it('THEN it should match the snapshot', () => {
        const { asFragment } = render(<TextInput {...props} />);
        expect(asFragment()).toMatchSnapshot();
      });

      it('THEN it should apply the error styles', () => {
        const input = screen.getByTestId(DATA_TEST.container);
        expect(input).toHaveClass('border-red-500');
      });
    });
  });
});
