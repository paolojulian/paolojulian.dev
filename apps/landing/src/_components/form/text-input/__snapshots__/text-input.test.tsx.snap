// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `[Function]`;

exports[`TESTING TextInput Component GIVEN a TextInput component with variant and isError prop WHEN the TextInput Component has isError set to true THEN it should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="relative w-full group transition-colors bg-inherit border-b border-opacity-20 hover:border-opacity-50 border-red-500"
    data-testid="text-input-container"
  >
    <label
      class="absolute left-3 top-0 transition duration-200 pointer-events-none font-medium text-sm text-new-highlight scale-0 opacity-0"
      data-testid="text-input-label"
      for="testing"
    >
      testing
    </label>
    <div
      class="absolute bottom-2 left-2 transition duration-200 pointer-events-none text-lg text-new-highlight scale-100 opacity-100"
    >
      testing
    </div>
    <input
      class="w-full px-3 pt-6 pb-2 bg-inherit focus:outline-none transition-opacity text-new-white opacity-0"
      data-testid="text-input-input"
      id="testing"
      name="testing"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`TESTING TextInput Component GIVEN a TextInput component with variant and isError prop WHEN the TextInput Component is rendered THEN it should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="relative w-full group transition-colors bg-inherit border-b border-opacity-20 hover:border-opacity-50 border-new-highlight text-new-white placeholder-new-highlight"
    data-testid="text-input-container"
  >
    <label
      class="absolute left-3 top-0 transition duration-200 pointer-events-none font-medium text-sm text-new-highlight scale-0 opacity-0"
      data-testid="text-input-label"
      for="testing"
    >
      testing
    </label>
    <div
      class="absolute bottom-2 left-2 transition duration-200 pointer-events-none text-lg text-new-highlight scale-100 opacity-100"
    >
      testing
    </div>
    <input
      class="w-full px-3 pt-6 pb-2 bg-inherit focus:outline-none transition-opacity text-new-white opacity-0"
      data-testid="text-input-input"
      id="testing"
      name="testing"
      value=""
    />
  </div>
</DocumentFragment>
`;
